### DISCUSSION TITLE
What is the default operation precision in MOOSE?

### DISCUSSION THREAD

**Discussion by abc-hy** [idaholab/moose#21799] [Q&A Modules: Phase field] [answered]
Hi everyone, 

What is the default operation precision in MOOSE?

How many digits after the decimal point?

Thank you.

**Comment by GiudGiud** [comment, marked as answer]
Hello

We use double precision for floating point numbers. 

> How many digits after the decimal point?

It's not that simple. Very grossly simplifying, 16, but please read more about this topic. (google `floating point precision`)

Guillaume

**Comment by ahzeeshan** [comment]
In my solution, I have one variable that can take values from 1e16 to 1 and another that can vary from 1e-33 to 1. How can I increase the moose precision during installation to accommodate these variables?

**Reply by GiudGiud** [reply]
hello

You'll need to start by modifying the petsc install to increase precision there
https://petsc.org/release/faq/#can-i-run-petsc-with-extended-precision

Then set LIBMESH_DEFAULT_QUADRUPLE_PRECISION when compiling libmesh. 

The MOOSE should adapt to the precision chosen in libmesh.

But this is not tested, so you'll need to adapt the code in a few places to get it to compile. It should not be too difficult however.

**Reply by GiudGiud** [reply]
I would first try to get the simulations working with the double precision. It's not because the variable values can vary over a large range that:
- all of the range is utilized
- even if all of the range is utilized, it can very well be that the matrix is still well conditioned and you get a workable solution

**Reply by ahzeeshan** [reply]
All of the variable range is utilized because the variable will have two values at the boundaries.

I have tried running it with default precision but get convergence problems. The code has a parameter V that can be varied - this parameter determines the range of values of the two variables. When V is low, the range of variables is smaller and I am able to get convergence. When I increase V, the solution does not converge. This is why I think it might have to do with precision.

**Reply by GiudGiud** [reply]
It could be that.

Or it could be a conditioning issue. See the troubleshooting failed solves instructions to see how to get the condition number of the matrix

**Reply by ahzeeshan** [reply]
Thanks for the suggestion. You are right, the condition number is very large - it becomes >1e11 after the first 3-4 non-linear iterations. I am using automatic scaling for all variables. Since scaling didn't help, is there any other approach to solve this issue? Will increasing precision help?

**Reply by GiudGiud** [reply]
I would try manual scaling as that can be more robust. 

Then you could split the domain with multiple variables that cover each different parts of the range of possible values, tied together using interface kernels, so that you can apply scaling independently to each part of the range
